ibm/service/power/resource_ibm_pi_network_security_group_rule.go

```diff
@@ -0,0 +1,611 @@
+		ruleMap[Attr_DestinationPort] = []map[string]interface{}{destinationPortMap}
+	}
+
+	ruleMap[Attr_ID] = rule.ID
+	ruleMap[Attr_Name] = rule.Name
+
+	protocolMap := networkSecurityGroupRuleProtocolToMap(rule.Protocol)
+	ruleMap[Attr_Protocol] = []map[string]interface{}{protocolMap}
+
+	remoteMap := networkSecurityGroupRuleRemoteToMap(rule.Remote)
+	ruleMap[Attr_Remote] = []map[string]interface{}{remoteMap}
+
+	if rule.SourcePort != nil {
+		sourcePortMap := networkSecurityGroupRulePortToMap(rule.SourcePort)
+		ruleMap[Attr_SourcePort] = []map[string]interface{}{sourcePortMap}
+	}
+
+	return ruleMap
+}
+
+func networkSecurityGroupRulePortToMap(port *models.NetworkSecurityGroupRulePort) map[string]interface{} {
+	portMap := make(map[string]interface{})
+	portMap[Attr_Maximum] = port.Maximum
+	portMap[Attr_Minimum] = port.Minimum
+	return portMap
+}
+
+func networkSecurityGroupRuleProtocolToMap(protocol *models.NetworkSecurityGroupRuleProtocol) map[string]interface{} {
+	protocolMap := make(map[string]interface{})
+	if protocol.IcmpTypes != nil {
+		protocolMap[Attr_ICMPTypes] = protocol.IcmpTypes
+	}
+	if protocol.TCPFlags != nil {
+		tcpFlags := []map[string]interface{}{}
+		for _, tcpFlagsItem := range protocol.TCPFlags {
+			tcpFlagsItemMap := make(map[string]interface{})
+			tcpFlagsItemMap[Attr_Flag] = tcpFlagsItem.Flag
+			tcpFlags = append(tcpFlags, tcpFlagsItemMap)
+		}
+		protocolMap[Attr_TCPFlags] = tcpFlags
+	}
+	if protocol.Type != "" {
+		protocolMap[Attr_Type] = protocol.Type
+	}
+	return protocolMap
+}
+
+func networkSecurityGroupRuleRemoteToMap(remote *models.NetworkSecurityGroupRuleRemote) map[string]interface{} {
+	remoteMap := make(map[string]interface{})
+	if remote.ID != "" {
+		remoteMap[Attr_ID] = remote.ID
+	}
+	if remote.Type != "" {
+		remoteMap[Attr_Type] = remote.Type
+	}
+	return remoteMap
+}
+
+func networkSecurityGroupRuleMapToPort(portMap map[string]interface{}) *models.NetworkSecurityGroupRulePort {
+	networkSecurityGroupRulePort := models.NetworkSecurityGroupRulePort{}
+	networkSecurityGroupRulePort.Maximum = portMap[Attr_Maximum].(float64)
+	networkSecurityGroupRulePort.Minimum = portMap[Attr_Minimum].(float64)
+	return &networkSecurityGroupRulePort
+}
+
+func networkSecurityGroupRuleMapToRemote(remoteMap map[string]interface{}) *models.NetworkSecurityGroupRuleRemote {
+	networkSecurityGroupRuleRemote := models.NetworkSecurityGroupRuleRemote{}
+	networkSecurityGroupRuleRemote.ID = remoteMap[Arg_ID].(string)
+	networkSecurityGroupRuleRemote.Type = remoteMap[Arg_Type].(string)
+	return &networkSecurityGroupRuleRemote
+}
+
+func networkSecurityGroupRuleMapToProtocol(protocolMap map[string]interface{}) *models.NetworkSecurityGroupRuleProtocol {
+	networkSecurityGroupRuleProtocol := models.NetworkSecurityGroupRuleProtocol{}
+	networkSecurityGroupRuleProtocol.Type = protocolMap[Arg_Type].(string)
+
+	if networkSecurityGroupRuleProtocol.Type == "icmp" {
```

Comment:
Could you add a const for these ? `"icmp", "tcp"`

Reply:
Updated